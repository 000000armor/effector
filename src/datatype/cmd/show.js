//@flow

import * as Name from './type'
import type {Cmd} from './impl'

export function show(value: Cmd): string {
 switch (value.type) {
  case Name.EMIT:
   return `Emit(${value.data.fullName})`
  case Name.RUN:
   return 'Run'
  case Name.COMPUTE:
   return 'Compute'
  case Name.FILTER:
   return 'Filter'
  case Name.UPDATE:
   return 'Update'
  default:
   /*::(value.type: empty)*/
   throw new Error('impossible type')
 }
}
