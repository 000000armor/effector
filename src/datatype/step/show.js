//@flow

import * as Name from './type'
import type {Step} from './Step.bs'

import {show as showCmd} from '../cmd/Cmd.bs'

export function show(value: Step): string {
 switch (value.type) {
  case Name.SINGLE:
   return `Single ${showCmd(value.data)}`
  case Name.MULTI: {
   if (value.data.size === 0) return 'Multi []'
   const inner = [...value.data]
    .map(e => move1(`  *  ${move2(show(e))}`))
    .join(`,\n`)
   return `Multi [\n${inner}\n]`
  }
  case Name.SEQ: {
   if (value.data.length === 0) return 'Seq []'
   const inner = value.data
    .map((e, n) => move1(`  ${n + 1}. ${move2(show(e))}`))
    .join(`,\n`)
   return `Seq [\n${inner}\n]`
  }
  default:
   /*::(value.type: empty)*/
   throw new Error('impossible type')
 }
}

const map1 = (s, i) => (i > 0 ? `  ${s}` : s)
const map2 = (s, i, l) => {
 if (i === 0) return s
 if (i === l.length - 1) return s
 return ` ${s}`
}
const move1 = e => splitJoin(e, map1)
const move2 = e => splitJoin(e, map2)
function splitJoin(e, pad) {
 return e
  .split(`\n`)
  .map(pad)
  .join(`\n`)
}
