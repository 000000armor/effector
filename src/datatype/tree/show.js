//@flow

import {cmd} from '../FullDatatype.bs'

import type {Tree} from './index'
const showCmd = cmd.show
export function show(value: Tree): string {
  const cmdStr = showCmd(value.node)
  const id = value.id
  switch (value.childs.size) {
    case 0:
      return `${id} ${cmdStr}`
    case 1: {
      const [inner] = [...value.childs]
      // const inner = value.data
      //  .map((e, n) => move1(`  ${n + 1}. ${move2(show(e))}`))
      //  .join(`,\n`)
      return `${id} ${cmdStr} â†’ ${show(inner)}`
    }
    default: {
      const inner = [...value.childs]
        .map(e => move1(`  *  ${move2(show(e))}`))
        .join(`,\n`)
      return `${id} ${cmdStr} [\n${inner}\n]`
    }
  }
}

const map1 = (s, i) => (i > 0 ? `  ${s}` : s)
const map2 = (s, i, l) => {
  if (i === 0) return s
  if (i === l.length - 1) return s
  return ` ${s}`
}
const move1 = e => splitJoin(e, map1)
const move2 = e => splitJoin(e, map2)
function splitJoin(e, pad) {
  return e
    .split(`\n`)
    .map(pad)
    .join(`\n`)
}
