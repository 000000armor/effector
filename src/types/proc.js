export function processFlow({errors}) {
  return errors.map(e => {
    const result = {}
    const extra = e.extra
      .map(e1 => e1.message[0])
      .slice(1)
      .map(processItem)
    const message = processItem(e.message[0])
    result.refs = extra
    result.message = message
    return result
  })
}
function processItem(e) {
  e = removeType(e)
  e = normalizeLoc(e)
  return e
}
function removeType(e) {
  const item = Object.assign({}, e)
  delete item.type
  return item
}
function normalizeLoc(e) {
  const item = Object.assign({}, e)
  delete item.path
  delete item.line
  delete item.endline
  delete item.start
  delete item.end
  item.source = item.loc.source
  item.type = item.loc.type
  const loc = {
    start: {line: item.loc.start.line, column: item.loc.start.column},
    end: {line: item.loc.end.line, column: item.loc.end.column},
  }
  item.loc = loc
  return item
}
