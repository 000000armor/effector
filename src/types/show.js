import {resolve, relative} from 'path'
import {readFile} from 'fs-extra'
import {codeFrameColumns} from '@babel/code-frame'

const root = resolve(__dirname, '../types')
const printRoot = resolve(__dirname, '../..')
function fromPrintRoot(source) {
  return relative(printRoot, resolve(root, source))
}

const files = {}
function flatMap(list, cb) {
  return list.reduce((acc, x) => acc.concat(cb(x)), [])
}

export async function generateReport(errors) {
  const results = []
  const getLines = ({loc: {start, end}}) => [start.line, end.line]
  const lineNumbersPad = Math.max(
    ...flatMap(errors, ({message, refs}) => [
      getLines(message),
      ...flatMap(refs, getLines),
    ]),
  ).toString().length

  for (const {message, refs} of errors) {
    results.push(await printMessage(message, false, true, lineNumbersPad))
    let lastSource = message.source
    for (const ref of refs) {
      const printHead = lastSource !== ref.source
      lastSource = ref.source
      results.push(await printMessage(ref, true, printHead, lineNumbersPad))
    }
    results.push('  ')
  }
  return results.join(`\n\n`)
}

async function readSource(source) {
  if (source in files) return files[source]
  const path = resolve(root, source)
  const result = await readFile(path, 'utf8')
  files[source] = result
  return result
}

async function printMessage(msg, isRef, printHead, lineNumbersPad) {
  if (msg.type === 'LibFile') return printLibRef(msg, isRef, lineNumbersPad)
  const source = await readSource(msg.source)
  const head = printHead ? `  ${fromPrintRoot(msg.source)}\n` : ''
  let frame
  if (isRef) {
    frame = printFrameRef(
      codeFrameColumns(source, msg.loc, {
        linesAbove: 2,
        linesBelow: 2,
        // message
      }),
      msg.descr,
      2,
      lineNumbersPad,
    )
  } else {
    const pad = addIndent('|  ', 4 + 1)
    const withPad = line => `${pad}${line}`
    let descr = msg.descr
    const acc = []
    const max = 80
    const trailMin = 15
    if (descr.includes('because ')) {
      const index = descr.indexOf('because ')
      splitLongLine(descr.slice(0, index), acc, max, trailMin)
      splitLongLine(descr.slice(index), acc, max, trailMin)
    } else {
      splitLongLine(descr, acc, max, trailMin)
    }
    acc.push('')
    descr = ['', ...acc.map(withPad)].join(`\n`)
    frame = codeFrameColumns(source, msg.loc, {
      linesAbove: 2,
      linesBelow: 2,
      message: descr,
    })
    const padInit =
      frame
        .split(`\n`)
        .find(line => line.startsWith('>'))
        .indexOf('|') -
      1 -
      2
    const offset = lineNumbersPad - padInit
    frame = addIndent(frame, offset)
      .split(`\n`)
      .map((line, i, list) => {
        const trimmed = line.trimStart()
        if (trimmed.startsWith('>'))
          return `> ${addIndent(trimmed.slice(2), offset)}`
        if (i >= 4 && i <= list.length - 3)
          return addIndent(trimmed, lineNumbersPad + 3)
        return line
      })
      .join(`\n`)
  }
  return `${head}${frame}`
}
function splitLongLine(line, acc, max, trailMin) {
  let currentLine = line
  while (currentLine.length > max) {
    let bound = max - 1
    let trail = -1
    let trailFound = false
    let lastSpace
    while (!trailFound) {
      lastSpace = currentLine.lastIndexOf(' ', bound)
      if (lastSpace === -1) {
        acc.push(currentLine)
        return
      }
      trail = currentLine.length - lastSpace
      if (trail < trailMin) {
        bound = Math.max(0, lastSpace - 1)
      } else {
        trailFound = true
      }
    }
    acc.push(currentLine.slice(0, lastSpace))
    currentLine = currentLine.slice(lastSpace + 1)
  }
  acc.push(currentLine)
}
async function printLibRef(msg, isRef, lineNumbersPad) {
  const filledSource = [
    Array(Math.max(0, msg.loc.start.line - 1))
      .fill('')
      .join(`\n`),
    msg.context,
    '',
  ].join(`\n`)
  const frame = printFrameRef(
    codeFrameColumns(filledSource, msg.loc, {
      linesAbove: 0,
      linesBelow: 0,
      // message
    }),
    msg.descr,
    0,
    lineNumbersPad,
  )
  return `   ${msg.source}\n${frame}`
}
function printFrameRef(frame, descr, padTop, lineNumbersPad) {
  let padNumbers = lineNumbersPad
  let offset = 0
  const result = frame
    .split(`\n`)
    .reverse()
    .map((line, index, arr) => {
      const i = arr.length - 1 - index
      if (i === padTop + 1) {
        padNumbers = line.indexOf('|') - 1 - 2
        offset = lineNumbersPad - padNumbers
        const index = line.lastIndexOf('^')
        return `${line.slice(0, index + 1)}^ ${descr}`
      }
      if (i === padTop) {
        return `> ${addIndent(line.slice(2), offset)}`
      }
      return line
    })
    .reverse()
    .join(`\n`)
  const offsetResult = addIndent(result, offset)
  return offsetResult
    .split(`\n`)
    .map((line, i) => {
      if (i === padTop) return line.slice(offset)
      return line
    })
    .join(`\n`)
}
function addIndent(str, n = 2) {
  const space = Array(n)
    .fill(' ')
    .join('')
  const splitted = str.split(/\n/gi)
  const padded = splitted.map(line => `${space}${line}`)
  return padded.join(`\n`)
}
