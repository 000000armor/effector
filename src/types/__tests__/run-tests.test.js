// @flow

//$off
import execa from 'execa'
//$off
// import {outputJSON} from 'fs-extra'
import {processFlow} from '../proc'
import {generateReport} from '../show'

jest.setTimeout(50000)

test('TypeScript', async() => {
  try {
    const data = await execa('npx', ['tsc', '-p', 'src/types'])
    expect(data).toMatchSnapshot('resolved')
  } catch (err) {
    const cleanedMessage = err.message
      .replace(/src\/types\//gm, '')
      .replace(/error TS\d+: /gm, '')
    expect(cleanedMessage).toMatchSnapshot('rejected')
  }
})

test('Flow', async() => {
  try {
    const data = await execa('npx', [
      'flow',
      'focus-check',
      '--strip-root',
      '--json',
      'src/types/types.test.js',
    ])
    expect(JSON.parse(data.stdout)).toMatchSnapshot('resolved')
  } catch (err) {
    const data = JSON.parse(
      err.message.substring(err.message.indexOf('\n') + 1).trim(),
    )
    const result = await generateReport(processFlow(data))

    expect(result).toMatchSnapshot('json messages')
  }
})
