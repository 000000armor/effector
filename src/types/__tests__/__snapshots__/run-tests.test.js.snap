// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Flow: json messages 1`] = `
Array [
  Array [
    "Cannot assign \`c\` to \`check2\` because number [1] is incompatible with string [2] in type argument \`Payload\` [3].",
  ],
  Array [
    "Cannot assign \`e\` to \`check4\` because number [1] is incompatible with string [2] in type argument \`State\` [3].",
  ],
  Array [
    "Cannot assign \`g\` to \`check6\` because string [1] is incompatible with number [2] in type argument \`Payload\` [3].",
  ],
  Array [
    "Cannot assign \`computed\` to \`check2\` because string [1] is incompatible with number [2] in type argument \`Payload\` [3].",
  ],
  Array [
    "Cannot assign \`createStore(...)\` to \`store2\` because number [1] is incompatible with string [2] in type argument \`State\` [3].",
  ],
  Array [
    "Cannot assign \`computed\` to \`check2\` because string [1] is incompatible with number [2] in type argument \`State\` [3].",
  ],
  Array [
    "Cannot call \`createDomain\` with \`234\` bound to \`domainName\` because number [1] is incompatible with string [2].",
  ],
  Array [
    "Cannot call \`createDomain\` with object literal bound to \`domainName\` because object literal [1] is incompatible with string [2].",
  ],
]
`;

exports[`Flow: json rejected 1`] = `
Object {
  "errors": Array [
    Object {
      "extra": Array [
        Object {
          "message": Array [
            Object {
              "context": null,
              "descr": "References:",
              "end": 0,
              "endline": 0,
              "line": 0,
              "path": "",
              "start": 1,
              "type": "Blame",
            },
          ],
        },
        Object {
          "message": Array [
            Object {
              "context": "    const a = createEvent<number>()",
              "descr": "[1]",
              "end": 32,
              "endline": 25,
              "line": 25,
              "loc": Object {
                "end": Object {
                  "column": 32,
                  "line": 25,
                  "offset": 389,
                },
                "source": "types.test.js",
                "start": Object {
                  "column": 27,
                  "line": 25,
                  "offset": 383,
                },
                "type": "SourceFile",
              },
              "path": "types.test.js",
              "start": 27,
              "type": "Blame",
            },
          ],
        },
        Object {
          "message": Array [
            Object {
              "context": "    const check2: Event<string> = c",
              "descr": "[2]",
              "end": 30,
              "endline": 30,
              "line": 30,
              "loc": Object {
                "end": Object {
                  "column": 30,
                  "line": 30,
                  "offset": 524,
                },
                "source": "types.test.js",
                "start": Object {
                  "column": 25,
                  "line": 30,
                  "offset": 518,
                },
                "type": "SourceFile",
              },
              "path": "types.test.js",
              "start": 25,
              "type": "Blame",
            },
          ],
        },
        Object {
          "message": Array [
            Object {
              "context": "declare export class Event<Payload> implements Unit<Payload> {",
              "descr": "[3]",
              "end": 34,
              "endline": 29,
              "line": 29,
              "loc": Object {
                "end": Object {
                  "column": 34,
                  "line": 29,
                  "offset": 480,
                },
                "source": "../../packages/effector/index.js.flow",
                "start": Object {
                  "column": 28,
                  "line": 29,
                  "offset": 473,
                },
                "type": "SourceFile",
              },
              "path": "../../packages/effector/index.js.flow",
              "start": 28,
              "type": "Blame",
            },
          ],
        },
      ],
      "kind": "infer",
      "level": "error",
      "message": Array [
        Object {
          "context": "    const check2: Event<string> = c",
          "descr": "Cannot assign \`c\` to \`check2\` because number [1] is incompatible with string [2] in type argument \`Payload\` [3].",
          "end": 35,
          "endline": 30,
          "line": 30,
          "loc": Object {
            "end": Object {
              "column": 35,
              "line": 30,
              "offset": 529,
            },
            "source": "types.test.js",
            "start": Object {
              "column": 35,
              "line": 30,
              "offset": 528,
            },
            "type": "SourceFile",
          },
          "path": "types.test.js",
          "start": 35,
          "type": "Blame",
        },
      ],
      "suppressions": Array [],
    },
    Object {
      "extra": Array [
        Object {
          "message": Array [
            Object {
              "context": null,
              "descr": "References:",
              "end": 0,
              "endline": 0,
              "line": 0,
              "path": "",
              "start": 1,
              "type": "Blame",
            },
          ],
        },
        Object {
          "message": Array [
            Object {
              "context": "    const check3: Store<number> = e",
              "descr": "[1]",
              "end": 30,
              "endline": 35,
              "line": 35,
              "loc": Object {
                "end": Object {
                  "column": 30,
                  "line": 35,
                  "offset": 618,
                },
                "source": "types.test.js",
                "start": Object {
                  "column": 25,
                  "line": 35,
                  "offset": 612,
                },
                "type": "SourceFile",
              },
              "path": "types.test.js",
              "start": 25,
              "type": "Blame",
            },
          ],
        },
        Object {
          "message": Array [
            Object {
              "context": "    const check4: Store<string> = e",
              "descr": "[2]",
              "end": 30,
              "endline": 36,
              "line": 36,
              "loc": Object {
                "end": Object {
                  "column": 30,
                  "line": 36,
                  "offset": 654,
                },
                "source": "types.test.js",
                "start": Object {
                  "column": 25,
                  "line": 36,
                  "offset": 648,
                },
                "type": "SourceFile",
              },
              "path": "types.test.js",
              "start": 25,
              "type": "Blame",
            },
          ],
        },
        Object {
          "message": Array [
            Object {
              "context": "declare export class Store<State> implements Unit<State> {",
              "descr": "[3]",
              "end": 32,
              "endline": 69,
              "line": 69,
              "loc": Object {
                "end": Object {
                  "column": 32,
                  "line": 69,
                  "offset": 1726,
                },
                "source": "../../packages/effector/index.js.flow",
                "start": Object {
                  "column": 28,
                  "line": 69,
                  "offset": 1721,
                },
                "type": "SourceFile",
              },
              "path": "../../packages/effector/index.js.flow",
              "start": 28,
              "type": "Blame",
            },
          ],
        },
      ],
      "kind": "infer",
      "level": "error",
      "message": Array [
        Object {
          "context": "    const check4: Store<string> = e",
          "descr": "Cannot assign \`e\` to \`check4\` because number [1] is incompatible with string [2] in type argument \`State\` [3].",
          "end": 35,
          "endline": 36,
          "line": 36,
          "loc": Object {
            "end": Object {
              "column": 35,
              "line": 36,
              "offset": 659,
            },
            "source": "types.test.js",
            "start": Object {
              "column": 35,
              "line": 36,
              "offset": 658,
            },
            "type": "SourceFile",
          },
          "path": "types.test.js",
          "start": 35,
          "type": "Blame",
        },
      ],
      "suppressions": Array [],
    },
    Object {
      "extra": Array [
        Object {
          "message": Array [
            Object {
              "context": null,
              "descr": "References:",
              "end": 0,
              "endline": 0,
              "line": 0,
              "path": "",
              "start": 1,
              "type": "Blame",
            },
          ],
        },
        Object {
          "message": Array [
            Object {
              "context": "    const f = createEffect<string, any, any>()",
              "descr": "[1]",
              "end": 33,
              "endline": 38,
              "line": 38,
              "loc": Object {
                "end": Object {
                  "column": 33,
                  "line": 38,
                  "offset": 694,
                },
                "source": "types.test.js",
                "start": Object {
                  "column": 28,
                  "line": 38,
                  "offset": 688,
                },
                "type": "SourceFile",
              },
              "path": "types.test.js",
              "start": 28,
              "type": "Blame",
            },
          ],
        },
        Object {
          "message": Array [
            Object {
              "context": "    const check6: Event<number> = g",
              "descr": "[2]",
              "end": 30,
              "endline": 42,
              "line": 42,
              "loc": Object {
                "end": Object {
                  "column": 30,
                  "line": 42,
                  "offset": 802,
                },
                "source": "types.test.js",
                "start": Object {
                  "column": 25,
                  "line": 42,
                  "offset": 796,
                },
                "type": "SourceFile",
              },
              "path": "types.test.js",
              "start": 25,
              "type": "Blame",
            },
          ],
        },
        Object {
          "message": Array [
            Object {
              "context": "declare export class Event<Payload> implements Unit<Payload> {",
              "descr": "[3]",
              "end": 34,
              "endline": 29,
              "line": 29,
              "loc": Object {
                "end": Object {
                  "column": 34,
                  "line": 29,
                  "offset": 480,
                },
                "source": "../../packages/effector/index.js.flow",
                "start": Object {
                  "column": 28,
                  "line": 29,
                  "offset": 473,
                },
                "type": "SourceFile",
              },
              "path": "../../packages/effector/index.js.flow",
              "start": 28,
              "type": "Blame",
            },
          ],
        },
      ],
      "kind": "infer",
      "level": "error",
      "message": Array [
        Object {
          "context": "    const check6: Event<number> = g",
          "descr": "Cannot assign \`g\` to \`check6\` because string [1] is incompatible with number [2] in type argument \`Payload\` [3].",
          "end": 35,
          "endline": 42,
          "line": 42,
          "loc": Object {
            "end": Object {
              "column": 35,
              "line": 42,
              "offset": 807,
            },
            "source": "types.test.js",
            "start": Object {
              "column": 35,
              "line": 42,
              "offset": 806,
            },
            "type": "SourceFile",
          },
          "path": "types.test.js",
          "start": 35,
          "type": "Blame",
        },
      ],
      "suppressions": Array [],
    },
    Object {
      "extra": Array [
        Object {
          "message": Array [
            Object {
              "context": null,
              "descr": "References:",
              "end": 0,
              "endline": 0,
              "line": 0,
              "path": "",
              "start": 1,
              "type": "Blame",
            },
          ],
        },
        Object {
          "message": Array [
            Object {
              "context": "    const check1: Event<string> = computed",
              "descr": "[1]",
              "end": 30,
              "endline": 54,
              "line": 54,
              "loc": Object {
                "end": Object {
                  "column": 30,
                  "line": 54,
                  "offset": 1071,
                },
                "source": "types.test.js",
                "start": Object {
                  "column": 25,
                  "line": 54,
                  "offset": 1065,
                },
                "type": "SourceFile",
              },
              "path": "types.test.js",
              "start": 25,
              "type": "Blame",
            },
          ],
        },
        Object {
          "message": Array [
            Object {
              "context": "    const check2: Event<number> = computed",
              "descr": "[2]",
              "end": 30,
              "endline": 55,
              "line": 55,
              "loc": Object {
                "end": Object {
                  "column": 30,
                  "line": 55,
                  "offset": 1114,
                },
                "source": "types.test.js",
                "start": Object {
                  "column": 25,
                  "line": 55,
                  "offset": 1108,
                },
                "type": "SourceFile",
              },
              "path": "types.test.js",
              "start": 25,
              "type": "Blame",
            },
          ],
        },
        Object {
          "message": Array [
            Object {
              "context": "declare export class Event<Payload> implements Unit<Payload> {",
              "descr": "[3]",
              "end": 34,
              "endline": 29,
              "line": 29,
              "loc": Object {
                "end": Object {
                  "column": 34,
                  "line": 29,
                  "offset": 480,
                },
                "source": "../../packages/effector/index.js.flow",
                "start": Object {
                  "column": 28,
                  "line": 29,
                  "offset": 473,
                },
                "type": "SourceFile",
              },
              "path": "../../packages/effector/index.js.flow",
              "start": 28,
              "type": "Blame",
            },
          ],
        },
      ],
      "kind": "infer",
      "level": "error",
      "message": Array [
        Object {
          "context": "    const check2: Event<number> = computed",
          "descr": "Cannot assign \`computed\` to \`check2\` because string [1] is incompatible with number [2] in type argument \`Payload\` [3].",
          "end": 42,
          "endline": 55,
          "line": 55,
          "loc": Object {
            "end": Object {
              "column": 42,
              "line": 55,
              "offset": 1126,
            },
            "source": "types.test.js",
            "start": Object {
              "column": 35,
              "line": 55,
              "offset": 1118,
            },
            "type": "SourceFile",
          },
          "path": "types.test.js",
          "start": 35,
          "type": "Blame",
        },
      ],
      "suppressions": Array [],
    },
    Object {
      "extra": Array [
        Object {
          "message": Array [
            Object {
              "context": null,
              "descr": "References:",
              "end": 0,
              "endline": 0,
              "line": 0,
              "path": "",
              "start": 1,
              "type": "Blame",
            },
          ],
        },
        Object {
          "message": Array [
            Object {
              "context": "    const store2: Store<string> = createStore(0)",
              "descr": "[1]",
              "end": 47,
              "endline": 70,
              "line": 70,
              "loc": Object {
                "end": Object {
                  "column": 47,
                  "line": 70,
                  "offset": 1444,
                },
                "source": "types.test.js",
                "start": Object {
                  "column": 47,
                  "line": 70,
                  "offset": 1443,
                },
                "type": "SourceFile",
              },
              "path": "types.test.js",
              "start": 47,
              "type": "Blame",
            },
          ],
        },
        Object {
          "message": Array [
            Object {
              "context": "    const store2: Store<string> = createStore(0)",
              "descr": "[2]",
              "end": 30,
              "endline": 70,
              "line": 70,
              "loc": Object {
                "end": Object {
                  "column": 30,
                  "line": 70,
                  "offset": 1427,
                },
                "source": "types.test.js",
                "start": Object {
                  "column": 25,
                  "line": 70,
                  "offset": 1421,
                },
                "type": "SourceFile",
              },
              "path": "types.test.js",
              "start": 25,
              "type": "Blame",
            },
          ],
        },
        Object {
          "message": Array [
            Object {
              "context": "declare export class Store<State> implements Unit<State> {",
              "descr": "[3]",
              "end": 32,
              "endline": 69,
              "line": 69,
              "loc": Object {
                "end": Object {
                  "column": 32,
                  "line": 69,
                  "offset": 1726,
                },
                "source": "../../packages/effector/index.js.flow",
                "start": Object {
                  "column": 28,
                  "line": 69,
                  "offset": 1721,
                },
                "type": "SourceFile",
              },
              "path": "../../packages/effector/index.js.flow",
              "start": 28,
              "type": "Blame",
            },
          ],
        },
      ],
      "kind": "infer",
      "level": "error",
      "message": Array [
        Object {
          "context": "    const store2: Store<string> = createStore(0)",
          "descr": "Cannot assign \`createStore(...)\` to \`store2\` because number [1] is incompatible with string [2] in type argument \`State\` [3].",
          "end": 47,
          "endline": 70,
          "line": 70,
          "loc": Object {
            "end": Object {
              "column": 47,
              "line": 70,
              "offset": 1444,
            },
            "source": "types.test.js",
            "start": Object {
              "column": 47,
              "line": 70,
              "offset": 1443,
            },
            "type": "SourceFile",
          },
          "path": "types.test.js",
          "start": 47,
          "type": "Blame",
        },
      ],
      "suppressions": Array [],
    },
    Object {
      "extra": Array [
        Object {
          "message": Array [
            Object {
              "context": null,
              "descr": "References:",
              "end": 0,
              "endline": 0,
              "line": 0,
              "path": "",
              "start": 1,
              "type": "Blame",
            },
          ],
        },
        Object {
          "message": Array [
            Object {
              "context": "    const check1: Store<string> = computed",
              "descr": "[1]",
              "end": 30,
              "endline": 119,
              "line": 119,
              "loc": Object {
                "end": Object {
                  "column": 30,
                  "line": 119,
                  "offset": 2775,
                },
                "source": "types.test.js",
                "start": Object {
                  "column": 25,
                  "line": 119,
                  "offset": 2769,
                },
                "type": "SourceFile",
              },
              "path": "types.test.js",
              "start": 25,
              "type": "Blame",
            },
          ],
        },
        Object {
          "message": Array [
            Object {
              "context": "    const check2: Store<number> = computed",
              "descr": "[2]",
              "end": 30,
              "endline": 121,
              "line": 121,
              "loc": Object {
                "end": Object {
                  "column": 30,
                  "line": 121,
                  "offset": 2819,
                },
                "source": "types.test.js",
                "start": Object {
                  "column": 25,
                  "line": 121,
                  "offset": 2813,
                },
                "type": "SourceFile",
              },
              "path": "types.test.js",
              "start": 25,
              "type": "Blame",
            },
          ],
        },
        Object {
          "message": Array [
            Object {
              "context": "declare export class Store<State> implements Unit<State> {",
              "descr": "[3]",
              "end": 32,
              "endline": 69,
              "line": 69,
              "loc": Object {
                "end": Object {
                  "column": 32,
                  "line": 69,
                  "offset": 1726,
                },
                "source": "../../packages/effector/index.js.flow",
                "start": Object {
                  "column": 28,
                  "line": 69,
                  "offset": 1721,
                },
                "type": "SourceFile",
              },
              "path": "../../packages/effector/index.js.flow",
              "start": 28,
              "type": "Blame",
            },
          ],
        },
      ],
      "kind": "infer",
      "level": "error",
      "message": Array [
        Object {
          "context": "    const check2: Store<number> = computed",
          "descr": "Cannot assign \`computed\` to \`check2\` because string [1] is incompatible with number [2] in type argument \`State\` [3].",
          "end": 42,
          "endline": 121,
          "line": 121,
          "loc": Object {
            "end": Object {
              "column": 42,
              "line": 121,
              "offset": 2831,
            },
            "source": "types.test.js",
            "start": Object {
              "column": 35,
              "line": 121,
              "offset": 2823,
            },
            "type": "SourceFile",
          },
          "path": "types.test.js",
          "start": 35,
          "type": "Blame",
        },
      ],
      "suppressions": Array [],
    },
    Object {
      "extra": Array [
        Object {
          "message": Array [
            Object {
              "context": null,
              "descr": "References:",
              "end": 0,
              "endline": 0,
              "line": 0,
              "path": "",
              "start": 1,
              "type": "Blame",
            },
          ],
        },
        Object {
          "message": Array [
            Object {
              "context": "    const domain3 = createDomain(234)",
              "descr": "[1]",
              "end": 36,
              "endline": 203,
              "line": 203,
              "loc": Object {
                "end": Object {
                  "column": 36,
                  "line": 203,
                  "offset": 4958,
                },
                "source": "types.test.js",
                "start": Object {
                  "column": 34,
                  "line": 203,
                  "offset": 4955,
                },
                "type": "SourceFile",
              },
              "path": "types.test.js",
              "start": 34,
              "type": "Blame",
            },
          ],
        },
        Object {
          "message": Array [
            Object {
              "context": "declare export function createDomain(domainName?: string): Domain",
              "descr": "[2]",
              "end": 56,
              "endline": 243,
              "line": 243,
              "loc": Object {
                "end": Object {
                  "column": 56,
                  "line": 243,
                  "offset": 6538,
                },
                "source": "../../packages/effector/index.js.flow",
                "start": Object {
                  "column": 51,
                  "line": 243,
                  "offset": 6532,
                },
                "type": "SourceFile",
              },
              "path": "../../packages/effector/index.js.flow",
              "start": 51,
              "type": "Blame",
            },
          ],
        },
      ],
      "kind": "infer",
      "level": "error",
      "message": Array [
        Object {
          "context": "    const domain3 = createDomain(234)",
          "descr": "Cannot call \`createDomain\` with \`234\` bound to \`domainName\` because number [1] is incompatible with string [2].",
          "end": 36,
          "endline": 203,
          "line": 203,
          "loc": Object {
            "end": Object {
              "column": 36,
              "line": 203,
              "offset": 4958,
            },
            "source": "types.test.js",
            "start": Object {
              "column": 34,
              "line": 203,
              "offset": 4955,
            },
            "type": "SourceFile",
          },
          "path": "types.test.js",
          "start": 34,
          "type": "Blame",
        },
      ],
      "suppressions": Array [],
    },
    Object {
      "extra": Array [
        Object {
          "message": Array [
            Object {
              "context": null,
              "descr": "References:",
              "end": 0,
              "endline": 0,
              "line": 0,
              "path": "",
              "start": 1,
              "type": "Blame",
            },
          ],
        },
        Object {
          "message": Array [
            Object {
              "context": "    const domain4 = createDomain({foo: true})",
              "descr": "[1]",
              "end": 44,
              "endline": 204,
              "line": 204,
              "loc": Object {
                "end": Object {
                  "column": 44,
                  "line": 204,
                  "offset": 5004,
                },
                "source": "types.test.js",
                "start": Object {
                  "column": 34,
                  "line": 204,
                  "offset": 4993,
                },
                "type": "SourceFile",
              },
              "path": "types.test.js",
              "start": 34,
              "type": "Blame",
            },
          ],
        },
        Object {
          "message": Array [
            Object {
              "context": "declare export function createDomain(domainName?: string): Domain",
              "descr": "[2]",
              "end": 56,
              "endline": 243,
              "line": 243,
              "loc": Object {
                "end": Object {
                  "column": 56,
                  "line": 243,
                  "offset": 6538,
                },
                "source": "../../packages/effector/index.js.flow",
                "start": Object {
                  "column": 51,
                  "line": 243,
                  "offset": 6532,
                },
                "type": "SourceFile",
              },
              "path": "../../packages/effector/index.js.flow",
              "start": 51,
              "type": "Blame",
            },
          ],
        },
      ],
      "kind": "infer",
      "level": "error",
      "message": Array [
        Object {
          "context": "    const domain4 = createDomain({foo: true})",
          "descr": "Cannot call \`createDomain\` with object literal bound to \`domainName\` because object literal [1] is incompatible with string [2].",
          "end": 44,
          "endline": 204,
          "line": 204,
          "loc": Object {
            "end": Object {
              "column": 44,
              "line": 204,
              "offset": 5004,
            },
            "source": "types.test.js",
            "start": Object {
              "column": 34,
              "line": 204,
              "offset": 4993,
            },
            "type": "SourceFile",
          },
          "path": "types.test.js",
          "start": 34,
          "type": "Blame",
        },
      ],
      "suppressions": Array [],
    },
  ],
  "flowVersion": "0.95.1",
  "jsonVersion": "1",
  "passed": false,
}
`;

exports[`TypeScript: rejected 1`] = `
"Command failed: npx tsc -p src/types

src/types/types.test.ts(30,11): error TS2322: Type 'Event<number>' is not assignable to type 'Event<string>'.
  Types of property 'watch' are incompatible.
    Type '(watcher: (payload: number) => any) => Subscription' is not assignable to type '(watcher: (payload: string) => any) => Subscription'.
      Types of parameters 'watcher' and 'watcher' are incompatible.
        Types of parameters 'payload' and 'payload' are incompatible.
          Type 'number' is not assignable to type 'string'.
src/types/types.test.ts(36,11): error TS2322: Type 'Store<number>' is not assignable to type 'Store<string>'.
  Type 'number' is not assignable to type 'string'.
src/types/types.test.ts(42,11): error TS2322: Type 'Event<string>' is not assignable to type 'Event<number>'.
  Types of property 'watch' are incompatible.
    Type '(watcher: (payload: string) => any) => Subscription' is not assignable to type '(watcher: (payload: number) => any) => Subscription'.
      Types of parameters 'watcher' and 'watcher' are incompatible.
        Types of parameters 'payload' and 'payload' are incompatible.
          Type 'string' is not assignable to type 'number'.
src/types/types.test.ts(55,11): error TS2322: Type 'Event<string>' is not assignable to type 'Event<number>'.
src/types/types.test.ts(70,11): error TS2322: Type 'Store<number>' is not assignable to type 'Store<string>'.
src/types/types.test.ts(121,11): error TS2322: Type 'Store<string>' is not assignable to type 'Store<number>'.
  Type 'string' is not assignable to type 'number'.
src/types/types.test.ts(203,34): error TS2345: Argument of type '234' is not assignable to parameter of type 'string | undefined'.
src/types/types.test.ts(204,34): error TS2345: Argument of type '{ foo: boolean; }' is not assignable to parameter of type 'string'.
"
`;
